src/TopOptEMFocus.jl

```diff
@@ -0,0 +1,550 @@
+
+
+# Then we create a function `gf_pf` that depends directly on $p_f$ and write out the derivate using adjoint method formula. Note that the threshold chainrule is already implemented in the functions above.
+# 
+
+function gf_pf(pf_vec; β, η, phys_params, fem_params)
+    pfh = FEFunction(fem_params.Pf, pf_vec)
+    pth = (pf -> Threshold(pf; β, η)) ∘ pfh
+    A_mat = MatrixA(pth; phys_params, fem_params)
+    b_vec = assemble_vector(v->(∫(v)fem_params.dΓ_s), fem_params.V)
+    u_vec = A_mat \ b_vec
+    
+    O_mat = MatrixOf(fem_params)
+    real(u_vec' * O_mat * u_vec)
+end
+
+function rrule(::typeof(gf_pf), pf_vec; β, η, phys_params, fem_params)
+    function U_pullback(dgdg)
+      NO_FIELDS, dgdg * Dgfdpf(pf_vec; β, η, phys_params, fem_params)
+    end
+    gf_pf(pf_vec; β, η, phys_params, fem_params), U_pullback
+end
+
+function Dgfdpf(pf_vec; β, η, phys_params, fem_params)
+    pfh = FEFunction(fem_params.Pf, pf_vec)
+    pth = (pf -> Threshold(pf; β, η)) ∘ pfh
+    A_mat = MatrixA(pth; phys_params, fem_params)
+    b_vec = assemble_vector(v->(∫(v)fem_params.dΓ_s), fem_params.V)
+    u_vec = A_mat \ b_vec
+    O_mat = MatrixOf(fem_params)
+    
+    uh = FEFunction(fem_params.U, u_vec)
+    w_vec =  A_mat' \ (O_mat * u_vec)
+    wconjh = FEFunction(fem_params.U, conj(w_vec)) 
+    
+    l_temp(dp) = ∫(real(-2 * DAdpf(uh, wconjh, pfh; phys_params, β, η)) * dp)fem_params.dΩ_d
+    dgfdpf = assemble_vector(l_temp, fem_params.Pf)
+    return dgfdpf
+end
+
+# Next we define the relation between $p_f$ and $p$, and get the derivative of the filter also using adjoint method.
+# 
+
+function pf_p0(p0; r, fem_params)
+    pf_vec = Filter(p0; r, fem_params)
+    pf_vec[pf_vec .< 0] .= 0
+    pf_vec[pf_vec .> 1] .= 1.0
+    pf_vec
+end
+
+function rrule(::typeof(pf_p0), p0; r, fem_params)
+  function pf_pullback(dgdpf)
+    NO_FIELDS, Dgdp(dgdpf; r, fem_params)
+  end
+  pf_p0(p0; r, fem_params), pf_pullback
+end
+
+function Dgdp(dgdpf; r, fem_params)
+    Af = assemble_matrix(fem_params.Pf, fem_params.Qf) do u, v
+        ∫(a_f(r, u, v))fem_params.dΩ_d + ∫(v * u)fem_params.dΩ_d
+    end
+    wvec = Af' \ dgdpf
+    wh = FEFunction(fem_params.Pf, wvec)
+    l_temp(dp) = ∫(wh * dp)fem_params.dΩ_d
+    return assemble_vector(l_temp, fem_params.P)
+end
+
+# Finally, we pack up everything and rewrite it as a compact form. We also record every optimization step values.
+# 
+
+function gf_p(p0::Vector; r, β, η, phys_params, fem_params)
+    pf_vec = pf_p0(p0; r, fem_params)
+    gf_pf(pf_vec; β, η, phys_params, fem_params)
+end
+
+function gf_p(p0::Vector, grad::Vector; r, β, η, phys_params, fem_params)
+    if length(grad) > 0
+        dgdp, = Zygote.gradient(p -> gf_p(p; r, β, η, phys_params, fem_params), p0)
+        grad[:] = dgdp
+    end
+    gvalue = gf_p(p0::Vector; r, β, η, phys_params, fem_params)
+    open("gvalue.txt", "a") do io
+        write(io, "$gvalue \n")
+    end
```

Comment:
I would suggest implementing this by passing a `logfile=nothing` keyword parameter and do:
```
if !isnothing(logfile)
    println(logfile, gvalue)
    flush(logfile)
end
```
Then, when you do the optimization, call `logfile = open("gvalue.txt", "w")` and pass it, then call `close(logfile)` when you are done.